Add iometer integration

jukrebs opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adds an integration for [IOmeter](https://iometer.de/) devices. 

IOmeter devices are used on the German electricity market to read digital electricity meters. This integration uses the local HTTP endpoint of the devices. This allows users to read live power, total energy usage, and total energy returned as well as health data of the device itself.

Python package and documentation:
https://github.com/iometer-gmbh/iometer.py
https://pypi.org/project/iometer/

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/36860

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

jukrebs committed:
add iometer integration

MaestroOnICe commented:
Hi all,

do I need to add more test to get codecov for all of the code? For now the integration is intended to be bronze and therefore only tests the config_flow.
Regards

jukrebs committed:
add zeroconf error test

jukrebs committed:
Merge branch 'home-assistant:dev' into maestroonice_iometer

abmantis commented:
> do I need to add more test to get codecov for all of the code? For now the integration is intended to be bronze and therefore only tests the config_flow.

If I am not mistaken only config_flow requires 100% test coverage, but it is highly recommended to have tests for the other parts of the code :)

By the way, ruff is failing.

abmantis commented on homeassistant/components/iometer/config_flow.py:
```diff
@@ -0,0 +1,94 @@
+        await self.async_set_unique_id(status.device.id)
+        self._abort_if_unique_id_configured()
+
+        self.context.update(
+            {
+                "title_placeholders": {
+                    "meter_number": self.data["meter_number"],
+                }
+            }
+        )
```
```suggestion
        self.context["title_placeholders"] = {
            "meter_number": self.data["meter_number"],
        }
```

abmantis replied:
Also, since the `flow_title` is not meant to be translated, could we set `name: f"IOmeter {self.data["meter_number"]"` instead and remove `flow_title` from strings.json?

MaestroOnICe replied:
Yes

abmantis commented on homeassistant/components/iometer/config_flow.py:
```diff
@@ -0,0 +1,94 @@
+        """Confirm discovery."""
+        if user_input is not None:
+            return self.async_create_entry(
+                title=f"IOmeter-{self.data['meter_number']}",
```
```suggestion
                title=f"IOmeter {self.data['meter_number']}",
```

abmantis commented on homeassistant/components/iometer/config_flow.py:
```diff
@@ -0,0 +1,94 @@
+                await self.async_set_unique_id(status.device.id)
+                self._abort_if_unique_id_configured()
+                return self.async_create_entry(
+                    title=f"IOmeter-{self.data['meter_number']}",
```
```suggestion
                    title=f"IOmeter {self.data['meter_number']}",
```
maybe we could to an internal `_async_create_entry(host, meter_number)` method that calls `self.async_create_entry` to prevent the title from diverging in the future?

MaestroOnICe replied:
Can you explain in what scenario the title would diverge?

MaestroOnICe replied:
Okay I understand. I agree, as this is done in many integrations.

abmantis commented on homeassistant/components/iometer/const.py:
```diff
@@ -0,0 +1,7 @@
+from typing import Final
+
+DOMAIN: Final = "iometer"
+DEFAULT_SCAN_INTERVAL: Final = timedelta(seconds=10)
```
`DEFAULT_SCAN_INTERVAL` is only used in the coordinator module. lets move it there

abmantis commented on homeassistant/components/iometer/coordinator.py:
```diff
@@ -0,0 +1,53 @@
+            reading = await self.client.get_current_reading()
+            status = await self.client.get_current_status()
+        except IOmeterConnectionError as error:
+            raise UpdateFailed(error) from error
```
```suggestion
            raise UpdateFailed(f"Error communicating with IOmeter: {error}")
```

MaestroOnICe replied:
Yes this is more verbose

abmantis commented on homeassistant/components/iometer/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "config_flow": true,
+  "dependencies": [],
+  "documentation": "https://www.home-assistant.io/integrations/iometer",
+  "homekit": {},
```
```suggestion
```

MaestroOnICe replied:
Makes sense to remove

abmantis commented on homeassistant/components/iometer/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "name": "IOmeter",
+  "codeowners": ["@MaestroOnICe"],
+  "config_flow": true,
+  "dependencies": [],
```
```suggestion
```

MaestroOnICe replied:
Makes sense to remove

abmantis commented on homeassistant/components/iometer/manifest.json:
[integration_type](https://developers.home-assistant.io/docs/creating_integration_manifest?_highlight=manifes#integration-type) is missing

abmantis commented on homeassistant/components/iometer/sensor.py:
```diff
@@ -0,0 +1,160 @@
+        key="connection_status",
+        translation_key="connection_status",
+        device_class=SensorDeviceClass.ENUM,
+        options=["connected", "disconnected"],
```
should this one be a binary_sensor instead? 
(if so, we should add it in a separate PR since we should have only 1 platform per PR)

MaestroOnICe replied:
Yes, we could do that. How would the EntityDescription look like?

MaestroOnICe replied:
Okay, I removed it and will add it in a future PR

abmantis commented on homeassistant/components/iometer/sensor.py:
```diff
@@ -0,0 +1,160 @@
+        value_fn=lambda data: data.status.device.core.connection_status,
+    ),
+    IOmeterEntityDescription(
+        key="wifi_rssi",
```
can we set `entity_registry_enabled_default=False` on this one?

MaestroOnICe replied:
Yes it is possible, but what is the benefit?

MaestroOnICe replied:
Lets also disable this, as discussed below.

abmantis commented on homeassistant/components/iometer/sensor.py:
```diff
@@ -0,0 +1,160 @@
+        value_fn=lambda data: data.status.device.bridge.rssi,
+    ),
+    IOmeterEntityDescription(
+        key="core_bridge_rssi",
```
can we set `entity_registry_enabled_default=False` on this one?

MaestroOnICe replied:
Same for this sensor, what would the benefit be?

MaestroOnICe replied:
No the RSSI's are not used a lot. Therefore lets disable them by default.

abmantis commented on homeassistant/components/iometer/sensor.py:
```diff
@@ -0,0 +1,160 @@
+        key="attachment_status",
+        translation_key="attachment_status",
+        device_class=SensorDeviceClass.ENUM,
+        options=["attached", "detached"],
```
should also be a binary_sensor

MaestroOnICe replied:
This shouldn't be a binary sensor as I missed to add "unknown" to the options

MaestroOnICe replied:
Although I just saw that the binary sensor also has the unknown state. How would we set `unknown` as a state for the binary sensor when the return value is `None` of `data.status.device.core.attachment_status`.

abmantis replied:
you can set the state of the binary sensor to `None` and it will be shown as `Unknown`. 
alternatively you can also set `_attr_available` to `False` and it is shown as unavailable (greyed out).

MaestroOnICe replied:
Okay, removed for now

abmantis commented on homeassistant/components/iometer/sensor.py:
```diff
@@ -0,0 +1,160 @@
+class IOmeterSensor(IOmeterEntity, SensorEntity):
+    """Defines a IOmeter sensor."""
+
+    _attr_has_entity_name = True
```
```suggestion
```
already in `IOmeterEntity`

abmantis commented on homeassistant/components/iometer/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  common-modules: done
+  config-flow-test-coverage: done
+  config-flow: done
+  dependency-transparency: done
```
I think that the [last point of this rule](https://developers.home-assistant.io/docs/core/integration-quality-scale/rules/dependency-transparency/) is not met yet:
> The version of the dependency published on PyPI should correspond to a tagged release in an open online repository.

MaestroOnICe replied:
The dependency is under the MIT license and it is available on PyPi [here](https://pypi.org/project/iometer/). The package is published to PyPi with a github action that has to be triggered manually, as on push to main is sometimes not wanted. When providing new documentation for instance. Is this a concern? Also I saw that a tagged release is missing. I am going to fix this.

MaestroOnICe replied:
Done

abmantis replied:
It seems that the tag was created on the wrong commit. 
The version on pypi comes from commit https://github.com/iometer-gmbh/iometer.py/commit/c72118906d42fcc412a4fab1d493ced4927a23e0 while the tag is on https://github.com/iometer-gmbh/iometer.py/commit/a7155627ade168983ad315218f75bd0d61f1f26d

Can you delete the tag and recreate it on the correct commit please?

MaestroOnICe replied:
Oh my bad, done.

abmantis replied:
thanks!

jukrebs committed:
Update manifest.json

I removed "dependencies" and "homekit" properties and added an "integration_type".

MaestroOnICe commented on homeassistant/components/iometer/__init__.py:
```diff
@@ -0,0 +1,40 @@
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```
I have tried with:
`await entry.runtime_data.client.close()`

This results in:
`WARNING (MainThread) [homeassistant.helpers.frame] Detected that integration 'iometer' closes the Home Assistant aiohttp session at homeassistant/components/iometer/__init__.py, line 40: await entry.runtime_data.client.close()`

Maybe it is okay this way because I am passing a session created by Home Assistant in line 22. I am not very familiar with how Home Assistant handles sessions.

abmantis replied:
Oh, I think the integration should never close the HA session. You need to remove that from `IOmeterClient.close()` or, since ``IOmeterClient.close()` does not do anything extra, remove that call from the exception handler in L27.

MaestroOnICe replied:
Done

MaestroOnICe commented on homeassistant/components/iometer/config_flow.py:
```diff
@@ -0,0 +1,94 @@
+
+    def __init__(self) -> None:
+        """Initialize the config flow."""
+        self.data: dict[str, Any] = {}
```
Yes I agree. This is easier to read

jukrebs committed:
fixup config_flow, add verbose error message

jukrebs committed:
move DEFAULT_SCAN_INTERVAL to coordinator

jukrebs committed:
remove duplicate _attr_has_entity_name

jukrebs committed:
add "unknown" state to sensors

jukrebs committed:
update title in config_flow test

NoRi2909 commented on homeassistant/components/iometer/strings.json:
```diff
@@ -0,0 +1,81 @@
+        "name": "Signal Strength Core/Bridge"
+      },
+      "wifi_rssi": {
+        "name": "Signal Strength Wifi"
```
The four entities above should all use sentence-casing so translators have a better indication of what words to treat as names. For example: "Signal strength Core/Bridge" would indicate that we keep "Core/Bridge" in translations.

The rest: "Total consumption", "Total production", "Signal strength Wi-Fi"  ("Wi-Fi" as name, too)

MaestroOnICe replied:
Done

NoRi2909 commented on homeassistant/components/iometer/strings.json:
```diff
@@ -0,0 +1,81 @@
+        "name": "Meter number"
+      },
+      "pin_status": {
+        "name": "Pin status",
```
-> "PIN status"

MaestroOnICe replied:
Done

abmantis commented on tests/components/iometer/test_config_flow.py:
```diff
@@ -0,0 +1,174 @@
+    assert result["reason"] == "cannot_connect"
+
+
+async def test_flow_errors(
```
```suggestion
async def test_user_flow_connection_error(
```

abmantis commented on tests/components/iometer/test_config_flow.py:
```diff
@@ -0,0 +1,174 @@
+    assert result["reason"] == "already_configured"
+
+
+async def test_zeroconf_flow_error(
```
```suggestion
async def test_zeroconf_flow_connection_error(
```

jukrebs committed:
remove close() function call in inti.py

jukrebs committed:
remove connection_status and attachment_status sensors

jukrebs committed:
add entity_registry_enabled_default to RSSI sensors

jukrebs committed:
fix grammar strings.json

jukrebs committed:
rework test_config_flow.py

jukrebs committed:
add internal _async_create_entry

abmantis commented on homeassistant/components/iometer/strings.json:
```diff
@@ -0,0 +1,81 @@
+      },
+      "battery_level": {
+        "name": "Battery level"
+      },
+      "connection_status": {
+        "name": "Connection Core/Bridge",
+        "state": {
+          "connected": "Connected",
+          "disconnected": "Disconnected",
+          "unknown": "Unknown"
```
```suggestion
```
this is not used anymore

abmantis commented on homeassistant/components/iometer/strings.json:
```diff
@@ -0,0 +1,81 @@
+  },
+  "entity": {
+    "sensor": {
+      "attachment_status": {
+        "name": "Connection to meter",
+        "state": {
+          "attached": "Attached",
+          "detached": "Detached",
+          "unknown": "Unknown"
+        }
+      },
```
```suggestion
```
this is not used anymore

abmantis commented on tests/components/iometer/test_config_flow.py:
```diff
@@ -0,0 +1,175 @@
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        user_input={CONF_HOST: "10.0.0.2"},
```
```suggestion
        user_input={CONF_HOST: IP_ADDRESS},
```
lets use the const in all the other ocorrences too

MaestroOnICe replied:
Done

abmantis commented on homeassistant/components/iometer/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+rules:
+  # Bronze
+  action-setup: done # does not apply
```
```suggestion
    status: exempt
    comment: |
      This integration does not provide additional actions.
```

abmantis commented on homeassistant/components/iometer/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  config-flow-test-coverage: done
+  config-flow: done
+  dependency-transparency: done
+  docs-actions: done # does not apply
```
set to exempt

MaestroOnICe replied:
Done

abmantis commented on homeassistant/components/iometer/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  docs-high-level-description: done
+  docs-installation-instructions: done
+  docs-removal-instructions: done
+  entity-event-setup: done
```
should this be `exempt`, since the integration does not register for any event?

MaestroOnICe replied:
Yes I marked it as such

MaestroOnICe replied:
Done

abmantis commented on homeassistant/components/iometer/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  docs-configuration-parameters: todo
+  docs-installation-parameters: todo
+  entity-unavailable: done
+  integration-owner: todo
```
```suggestion
  integration-owner: done
```

MaestroOnICe replied:
Done

abmantis commented on homeassistant/components/iometer/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  integration-owner: todo
+  log-when-unavailable: done
+  parallel-updates: todo
+  reauthentication-flow: todo
```
I don't see any auth currently

MaestroOnICe replied:
Yes, and and there is none needed since it is a local HTTP endpoint.

abmantis replied:
Yes, it should be exempt if it has no auth

abmantis commented on homeassistant/components/iometer/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  devices: todo
+  diagnostics: todo
+  discovery-update-info: todo
+  discovery: todo
```
isn't this one `done`?

MaestroOnICe replied:
Yes it is.

MaestroOnICe replied:
Yes, marked as such.

abmantis commented on homeassistant/components/iometer/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  dynamic-devices: todo
+  entity-category: todo
+  entity-device-class: todo
+  entity-disabled-by-default: todo
```
```suggestion
  entity-disabled-by-default: done
```

MaestroOnICe replied:
Done

abmantis commented on homeassistant/components/iometer/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  docs-use-cases: todo
+  dynamic-devices: todo
+  entity-category: todo
+  entity-device-class: todo
```
```suggestion
  entity-device-class: done
```

MaestroOnICe replied:
Done

abmantis commented on homeassistant/components/iometer/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  docs-troubleshooting: todo
+  docs-use-cases: todo
+  dynamic-devices: todo
+  entity-category: todo
```
```suggestion
  entity-category: done
```

MaestroOnICe replied:
Done

abmantis commented on homeassistant/components/iometer/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  entity-category: todo
+  entity-device-class: todo
+  entity-disabled-by-default: todo
+  entity-translations: todo
```
```suggestion
  entity-translations: done
```

abmantis commented on homeassistant/components/iometer/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+
+  # Platinum
+  async-dependency: todo
+  inject-websession: todo
```
```suggestion
  inject-websession: done
```

MaestroOnICe replied:
Done

abmantis commented on homeassistant/components/iometer/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  stale-devices: todo
+
+  # Platinum
+  async-dependency: todo
```
isn't the dependency async already?

MaestroOnICe replied:
Yes it is, I haven't looked this far done in the quality_scale.yml

MaestroOnICe commented on homeassistant/components/iometer/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  # Silver
+  action-exceptions: todo # does not apply
+  config-entry-unloading: done
+  docs-configuration-parameters: todo
```
Yes

MaestroOnICe commented on homeassistant/components/iometer/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  unique-config-entry: done
+
+  # Silver
+  action-exceptions: todo # does not apply
```
Done

MaestroOnICe commented on homeassistant/components/iometer/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  action-exceptions: todo # does not apply
+  config-entry-unloading: done
+  docs-configuration-parameters: todo
+  docs-installation-parameters: todo
```
Can be marked as done.

jukrebs committed:
update quality_scale

MaestroOnICe commented on homeassistant/components/iometer/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  log-when-unavailable: done
+  parallel-updates: todo
+  reauthentication-flow: todo
+  test-coverage: todo
```
I have a 97% coverage locally. Can we set this to done?

abmantis replied:
`__init__.py` and `coordinator.py` have less than 95% coverage, which this rule requires.

MaestroOnICe replied:
Oh I understood the requirement wrong. Lets keep it as `todo` for now, as I don't have much experience with implementing tests for Home Assistant and therefore have to look at this in more depth.

jukrebs committed:
fix strings.json and test_config_flow.py

abmantis commented on homeassistant/components/iometer/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  entity-unavailable: done
+  integration-owner: todo
+  log-when-unavailable: done
+  parallel-updates: todo
```
Yes, the reasoning is the same as the Autarco integration, for example.

MaestroOnICe replied:
Done

jukrebs committed:
fix quality_scale.yaml

abmantis reviewed (approved):
Thank you!
